![Why is sleep important](https://cdn-prod.medicalnewstoday.com/content/images/articles/325/325353/why-is-sleep-important.jpg)

*Image source: [Medical News Today — Why sleep is essential for health](https://www.medicalnewstoday.com/articles/325353)*

# Sleep health and lifestyle dataset analyze

In [43]:
import pandas as pd

df = pd.read_csv("Sleep_health_and_lifestyle_dataset.csv")
print(df.shape)
df.head(200)

(374, 13)


,Person ID,Gender,Age,Occupation,Sleep Duration,Quality of Sleep,Physical Activity Level,Stress Level,BMI Category,Blood Pressure,Heart Rate,Daily Steps,Sleep Disorder
0,1,Male,27,Software Engineer,6.1,6,42,6,Overweight,126/83,77,4200,NaN
1,2,Male,28,Doctor,6.2,6,60,8,Normal,125/80,75,10000,NaN
2,3,Male,28,Doctor,6.2,6,60,8,Normal,125/80,75,10000,NaN
3,4,Male,28,Sales Representative,5.9,4,30,8,Obese,140/90,85,3000,Sleep Apnea
4,5,Male,28,Sales Representative,5.9,4,30,8,Obese,140/90,85,3000,Sleep Apnea
...,...,...,...,...,...,...,...,...,...,...,...,...,...
195,196,Male,43,Salesperson,6.5,6,45,7,Overweight,130/85,72,6000,Insomnia
196,197,Male,43,Salesperson,6.5,6,45,7,Overweight,130/85,72,6000,Insomnia
197,198,Male,43,Salesperson,6.5,6,45,7,Overweight,130/85,72,6000,Insomnia
198,199,Male,43,Salesperson,6.5,6,45,7,Overweight,130/85,72,6000,Insomnia


### Which gender sleep longer?

In [21]:
avg_sleep = df.groupby("Gender")["Sleep Duration"].mean().round(2)
print(avg_sleep)

winner = avg_sleep.idxmax()
print(f"\n{winner}s sleep longer on average: {avg_sleep.max()} hours")

Gender
Female    7.23
Male      7.04
Name: Sleep Duration, dtype: float64

Females sleep longer on average: 7.23 hours


**Females sleep longer on average: 7.23 hours**

### Which gender has better sleep quality?

In [22]:
# Which gender has better sleep quality?
avg_quality = df.groupby("Gender")["Quality of Sleep"].mean().round(2)
print(avg_quality)

best = avg_quality.idxmax()
print(f"\n{best}s have better sleep quality on average: {avg_quality.max()} / 10")

Gender
Female    7.66
Male      6.97
Name: Quality of Sleep, dtype: float64

Females have better sleep quality on average: 7.66 / 10


**Females have better sleep quality on average: 7.66 / 10**

### Which occupation sleeps longer and shorter?

In [23]:
# Which occupation sleeps longer and shorter?
by_job = df.groupby("Occupation")["Sleep Duration"].mean().round(2).sort_values(ascending=False)
print(by_job)

print(f"\nLongest sleepers:  {by_job.idxmax()} ({by_job.max()} hours)")
print(f"Shortest sleepers: {by_job.idxmin()} ({by_job.min()} hours)")

Occupation
Engineer                7.99
Lawyer                  7.41
Accountant              7.11
Nurse                   7.06
Doctor                  6.97
Manager                 6.90
Software Engineer       6.75
Teacher                 6.69
Salesperson             6.40
Scientist               6.00
Sales Representative    5.90
Name: Sleep Duration, dtype: float64

Longest sleepers:  Engineer (7.99 hours)
Shortest sleepers: Sales Representative (5.9 hours)


### Which age group sleeps longer and shorter?

In [24]:
# Which age group sleeps longer and shorter?
df["Age Group"] = pd.cut(df["Age"], bins=[20, 30, 40, 50, 60],
                         labels=["20s", "30s", "40s", "50s"], right=False)

by_age = df.groupby("Age Group", observed=True)["Sleep Duration"].mean().round(2).sort_values(ascending=False)
print(by_age)

print(f"\nLongest sleepers:  people in their {by_age.idxmax()} ({by_age.max()} hours)")
print(f"Shortest sleepers: people in their {by_age.idxmin()} ({by_age.min()} hours)")

Age Group
50s    7.63
30s    7.07
40s    6.91
20s    6.47
Name: Sleep Duration, dtype: float64

Longest sleepers:  people in their 50s (7.63 hours)
Shortest sleepers: people in their 20s (6.47 hours)


### Correlation table between all the health / lifestyle numbers

In [27]:
# Correlation table between all the health / lifestyle numbers
num_cols = ["Age", "Sleep Duration", "Quality of Sleep",
            "Physical Activity Level", "Stress Level", "Heart Rate", "Daily Steps"]

corr = df[num_cols].corr().round(2)
corr  # display as a table (closer to +1 = move together, closer to -1 = opposite)

,Age,Sleep Duration,Quality of Sleep,Physical Activity Level,Stress Level,Heart Rate,Daily Steps
Age,1.00,0.34,0.47,0.18,-0.42,-0.23,0.06
Sleep Duration,0.34,1.00,0.88,0.21,-0.81,-0.52,-0.04
Quality of Sleep,0.47,0.88,1.00,0.19,-0.90,-0.66,0.02
Physical Activity Level,0.18,0.21,0.19,1.00,-0.03,0.14,0.77
Stress Level,-0.42,-0.81,-0.90,-0.03,1.00,0.67,0.19
Heart Rate,-0.23,-0.52,-0.66,0.14,0.67,1.00,-0.03
Daily Steps,0.06,-0.04,0.02,0.77,0.19,-0.03,1.00


### Occupation is a category, so we compare AVERAGES of each health metric per major

In [28]:
# Occupation is a category, so we compare AVERAGES of each health metric per major
by_occupation = df.groupby("Occupation")[num_cols].mean().round(1)
by_occupation = by_occupation.sort_values("Quality of Sleep", ascending=False)
by_occupation

,Age,Sleep Duration,Quality of Sleep,Physical Activity Level,Stress Level,Heart Rate,Daily Steps
Occupation,,,,,,,
Engineer,46.6,8.0,8.4,51.9,3.9,67.2,5981.0
Accountant,39.6,7.1,7.9,58.1,4.6,68.9,6881.1
Lawyer,39.4,7.4,7.9,70.4,5.1,69.6,7661.7
Nurse,51.8,7.1,7.4,78.6,5.5,72.0,8057.5
Manager,45.0,6.9,7.0,55.0,5.0,75.0,5500.0
Teacher,41.7,6.7,7.0,45.6,4.5,67.2,5957.5
Doctor,32.7,7.0,6.6,55.4,6.7,71.5,6808.5
Software Engineer,31.2,6.8,6.5,48.0,6.0,75.5,5800.0
Salesperson,43.5,6.4,6.0,45.0,7.0,72.0,6000.0


**Sleep Dutation** is longest for **Engineers** **8 hours**

**Sleep Duration** is shortest for **Sales Representative 5.9 hours**

**The Most** Physical Active is **Nurse 78.6** 

**The Least** Physical Active major is **Sales Representative 30** 

**Nurse** do most **daily steps 8057.5**

**Sales Representative** do least **daily steps 3000**

### BMI category

In [34]:
# All BMI categories and how many people are in each
bmi_counts = df["BMI Category"].value_counts()
print(bmi_counts)

BMI Category
Normal           195
Overweight       148
Normal Weight     21
Obese             10
Name: count, dtype: int64


In [35]:
# Which major tends to have which BMI category?
# (merge the duplicate "Normal Weight" label into "Normal" first)
df["BMI Category"] = df["BMI Category"].replace({"Normal Weight": "Normal"})

# Row % = of each occupation, what share falls in each BMI category
bmi_by_job = (pd.crosstab(df["Occupation"], df["BMI Category"], normalize="index") * 100).round(0)
print(bmi_by_job)

# The BMI category each major tends to have most
print("\nMost common BMI per occupation:")
print(bmi_by_job.idxmax(axis=1))

BMI Category          Normal  Obese  Overweight
Occupation                                     
Accountant              84.0    0.0        16.0
Doctor                  94.0    6.0         0.0
Engineer                95.0    0.0         5.0
Lawyer                  91.0    4.0         4.0
Manager                  0.0    0.0       100.0
Nurse                   10.0    0.0        90.0
Sales Representative     0.0  100.0         0.0
Salesperson              0.0    0.0       100.0
Scientist                0.0    0.0       100.0
Software Engineer       50.0   25.0        25.0
Teacher                 15.0    2.0        82.0

Most common BMI per occupation:
Occupation
Accountant                  Normal
Doctor                      Normal
Engineer                    Normal
Lawyer                      Normal
Manager                 Overweight
Nurse                   Overweight
Sales Representative         Obese
Salesperson             Overweight
Scientist               Overweight
Software Engineer

### Which gender tends to have insomnia?

In [36]:
# Which gender tends to have insomnia?
# Count of insomnia cases per gender
insomnia = df[df["Sleep Disorder"] == "Insomnia"]
insomnia_count = insomnia["Gender"].value_counts()
print("Number of insomnia cases per gender:")
print(insomnia_count)

# Share = of each gender, what % suffer from insomnia
insomnia_rate = (df["Sleep Disorder"].eq("Insomnia")
                 .groupby(df["Gender"]).mean() * 100).round(1)
print("\n% of each gender that has insomnia:")
print(insomnia_rate)

winner = insomnia_rate.idxmax()
print(f"\n{winner}s tend to have insomnia more: {insomnia_rate.max()}% of them, "
      f"vs {insomnia_rate.min()}% of {insomnia_rate.idxmin()}s")

Number of insomnia cases per gender:
Gender
Male      41
Female    36
Name: count, dtype: int64

% of each gender that has insomnia:
Gender
Female    19.5
Male      21.7
Name: Sleep Disorder, dtype: float64

Males tend to have insomnia more: 21.7% of them, vs 19.5% of Females


**Males tend to have insomnia more: 21.7% of them have it (41 cases), vs 19.5% of Females (36 cases).**

### Correlation between blood pressure and heart rate

In [41]:
# Correlation between blood pressure and heart rate
# Blood Pressure is text like "126/83", so split it into two numbers first
df[["Systolic", "Diastolic"]] = df["Blood Pressure"].str.split("/", expand=True).astype(int)

bp_hr_corr = df[["Systolic", "Diastolic", "Heart Rate"]].corr().round(2)
print(bp_hr_corr)

print(f"\nSystolic  vs Heart Rate: {bp_hr_corr.loc['Systolic', 'Heart Rate']}")
print(f"Diastolic vs Heart Rate: {bp_hr_corr.loc['Diastolic', 'Heart Rate']}")
print("(closer to +1 = rise together, closer to -1 = opposite, near 0 = no link)")

            Systolic  Diastolic  Heart Rate
Systolic        1.00       0.97        0.29
Diastolic       0.97       1.00        0.27
Heart Rate      0.29       0.27        1.00

Systolic  vs Heart Rate: 0.29
Diastolic vs Heart Rate: 0.27
(closer to +1 = rise together, closer to -1 = opposite, near 0 = no link)


**Blood pressure and heart rate have only a weak positive correlation: Systolic vs Heart Rate = 0.29, Diastolic vs Heart Rate = 0.27.** They tend to rise together slightly, but the link is weak. (Systolic and Diastolic, on the other hand, are almost perfectly correlated at 0.97.)

### Correlation between stress and sleep quality

In [ ]:
# Correlation between stress and sleep quality
stress_sleep_corr = df["Stress Level"].corr(df["Quality of Sleep"]).round(2)
print(f"Stress Level vs Quality of Sleep: {stress_sleep_corr}")
print("(closer to -1 = strong opposite link: more stress -> worse sleep)")

**Stress and sleep quality have a strong negative correlation: -0.90.** This is the strongest relationship in the whole dataset — the more stressed a person is, the worse their sleep quality. Stress is the biggest enemy of good sleep here.

## 📊 Summary — What this data tells us about sleep

After exploring **374 people** and their sleep, health, and lifestyle, a clear story comes out:

### 😟 Stress is the #1 enemy of sleep
The strongest relationship in the entire dataset is between **stress and sleep quality: −0.90**. That's almost a perfect opposite link — the more stressed a person is, the worse they sleep. Stress also drags down sleep *duration* (−0.81) and raises *heart rate* (+0.67).

### 😴 Good sleep comes as a package
People who sleep **longer** also sleep **better** (+0.88). Sleep length and sleep quality go hand in hand — you rarely get one without the other.

### 👩 vs 👨 Gender differences
- **Women** sleep a bit longer (7.23h vs 7.04h) and rate their sleep higher (7.66 vs 6.97).
- But **men** are slightly more likely to suffer from **insomnia** (21.7% vs 19.5%).

### 💼 Your job affects your sleep
- **Engineers** sleep best of all (8.4 / 10), are the least stressed, and stay mostly at a normal weight.
- **Sales Representatives** sleep the worst (4 / 10), are the most stressed, take the fewest steps (3,000/day), and 100% fall into the obese category.

### ❤️ A surprise: blood pressure ≠ heart rate
We might expect blood pressure and heart rate to rise together, but the link is **weak (0.29)**. Knowing one tells you little about the other in this group.

### 🏃 Movement matters
More physically active people also walk more (steps vs activity = +0.77) — the two reinforce each other.

---

**One-line takeaway:**
> *Lower your stress and you'll sleep longer **and** better — and habits like your job and daily movement quietly shape your sleep just as much.*